massa-execution-worker/src/controller.rs

```diff
@@ -0,0 +1,190 @@
+pub struct ExecutionControllerImpl {
+    /// execution config
+    pub(crate) config: ExecutionConfig,
+    /// input data to process in the VM loop
+    /// with a wakeup condition variable that needs to be triggered when the data changes
+    pub(crate) input_data: Arc<(Condvar, Mutex<VMInputData>)>,
+    /// current execution state (see execution.rs for details)
+    pub(crate) execution_state: Arc<RwLock<ExecutionState>>,
+}
+
+impl ExecutionControllerImpl {
+    /// consumes and returns the input fed to the controller
+    pub(crate) fn consume_input(&mut self) -> VMInputData {
+        std::mem::take(&mut self.input_data.1.lock())
+    }
+}
+
+impl ExecutionController for ExecutionControllerImpl {
+    /// called to signal changes on the current blockclique, also listing newly finalized blocks
+    ///
+    /// # arguments
+    /// * finalized_blocks: list of newly finalized blocks to be appended to the input finalized blocks
+    /// * blockclique: new blockclique, replaces the curren one in the input
+    fn update_blockclique_status(
+        &self,
+        finalized_blocks: Map<BlockId, Block>,
+        blockclique: Map<BlockId, Block>,
+    ) {
+        // index newly finalized blocks by slot
+        let mapped_finalized_blocks: HashMap<_, _> = finalized_blocks
+            .into_iter()
+            .map(|(b_id, b)| (b.header.content.slot, (b_id, b)))
+            .collect();
+        // index blockclique by slot
+        let mapped_blockclique = blockclique
+            .into_iter()
+            .map(|(b_id, b)| (b.header.content.slot, (b_id, b)))
+            .collect();
+        //update input data
+        let mut input_data = self.input_data.1.lock();
+        input_data.blockclique = mapped_blockclique; // replace blockclique
+        input_data.finalized_blocks.extend(mapped_finalized_blocks); // append finalized blocks
+        input_data.blockclique_changed = true; // signal a blockclique change
+        self.input_data.0.notify_one(); // wake up VM loop
+    }
+
+    /// Get the generated execution events, optionnally filtered by:
+    /// * start slot
+    /// * end slot
+    /// * emitter address
+    /// * original caller address
+    /// * operation id
+    fn get_filtered_sc_output_event(
+        &self,
+        start: Option<Slot>,
+        end: Option<Slot>,
+        emitter_address: Option<Address>,
+        original_caller_address: Option<Address>,
+        original_operation_id: Option<OperationId>,
+    ) -> Vec<SCOutputEvent> {
+        self.execution_state.read().get_filtered_sc_output_event(
+            start,
+            end,
+            emitter_address,
+            original_caller_address,
+            original_operation_id,
+        )
+    }
+
+    /// gets a copy of a full ledger entry
+    ///
+    /// # return value
+    /// * (final_entry, active_entry)
+    fn get_full_ledger_entry(&self, addr: &Address) -> (Option<LedgerEntry>, Option<LedgerEntry>) {
```

Comment:
the general idea is to have a guarantee that renamed. But the two need to stay together for atomicity (taking the 2 values at the same moment) and because Active depends on the value of Final (see implementation)